File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    public int solution(int k, int[][] dungeons) {
        int answer = 0;
        int n = dungeons.length;
        int[] order = new int[n];
        for (int i = 0; i < n; i++) {
            order[i] = i;
        }
        do {
            int fatigue = k;
            int count = 0;
            for (int i : order) {
                if (fatigue >= dungeons[i][0]) {
                    fatigue -= dungeons[i][1];
                    count++;
                } else {
                    break;
                }
            }
            answer = Math.max(answer, count);
        } while (nextPermutation(order));
        return answer;
    }
    
    private boolean nextPermutation(int[] order) {
        int i = order.length - 1;
        while (i > 0 && order[i-1] >= order[i]) {
            i--;
        }
        if (i <= 0) {
            return false;
        }
        int j = order.length - 1;
        while (order[j] <= order[i-1]) {
            j--;
        }
        swap(order, i-1, j);
        j = order.length - 1;
        while (i < j) {
            swap(order, i, j);
            i++;
            j--;
        }
        return true;
    }
    
    private void swap(int[] order, int i, int j) {
        int temp = order[i];
        order[i] = order[j];
        order[j] = temp;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.swift
import Foundation

func solution(_ k:Int, _ dungeons:[[Int]]) -> Int {
    var maxCount = 0
    let n = dungeons.count
    var visited = [Bool](repeating: false, count: n)
    
    func dfs (_ fatigue: Int, _ count: Int){
        maxCount = max(maxCount, count)
        for i in 0..<n {
            if !visited[i] && fatigue >= dungeons[i][0]{
                visited[i] = true
                dfs(fatigue - dungeons[i][1], count + 1)
                visited[i] = false
            }
        }
    }
    dfs(k, 0)
    return maxCount
}